src/Adapter/Order/OrderProductUpdater.php

```diff
@@ -0,0 +1,519 @@
+
+            if ($delete) {
+                $orderDetail->delete();
+            }
+
+            StockAvailable::synchronize($productId);
+        } elseif ($orderDetail->id_warehouse == 0) {
+            StockAvailable::updateQuantity(
+                $orderDetail->product_id,
+                $orderDetail->product_attribute_id,
+                $quantityToReinject,
+                $orderDetail->id_shop,
+                true,
+                [
+                    'id_order' => $orderDetail->id_order,
+                    'id_stock_mvt_reason' => Configuration::get('PS_STOCK_CUSTOMER_RETURN_REASON'),
+                ]
+            );
+
+            // sync all stock
+            (new StockManager())->updatePhysicalProductQuantity(
+                (int) $orderDetail->id_shop,
+                (int) Configuration::get('PS_OS_ERROR'),
+                (int) Configuration::get('PS_OS_CANCELED'),
+                null,
+                (int) $orderDetail->id_order
+            );
+
+            if ($delete) {
+                $orderDetail->delete();
+            }
+        } else {
+            throw new OrderException('This product cannot be re-stocked.');
+        }
+    }
+
+    /**
+     * @param Cart $cart
+     * @param Order $order
+     * @param bool $delete
+     *
+     * @return Cart
+     *
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    private function syncCartAndOrderCartRules(Cart $cart, Order $order, bool $delete = false): Cart
+    {
+        Context::getContext()->cart = $cart;
+        CartRule::autoAddToCart();
+        CartRule::autoRemoveFromCart();
+
+        $orderCartRulesData = $order->getCartRules();
+        $orderCartRulesIds = array_map(
+            function (array $orderCartRuleData) { return (int) $orderCartRuleData['id_cart_rule']; },
+            $orderCartRulesData
+        );
+
+        $cartRules = $cart->getCartRules();
+        $cartRulesIds = array_map(
+            function (array $cartRule) { return (int) $cartRule['id_cart_rule']; },
+            $cartRules
+        );
+
+        if (count($orderCartRulesIds) === count($cartRulesIds)) {
+            return $cart;
+        }
+
+        if (count($orderCartRulesIds) > count($cartRulesIds)) {
+            if ($delete) { // Product deletion, the extra order cart rules have to be deleted
+                foreach ($order->getCartRules() as $orderCartRule) {
+                    foreach (array_diff($orderCartRulesIds, $cartRulesIds) as $idCartRule) {
+                        if ($idCartRule == $orderCartRule['id_cart_rule']) {
+                            $orderCartRule = new OrderCartRule($orderCartRule['id_order_cart_rule']);
+                            $orderCartRule->delete();
+                        }
+                    }
+                }
+            } else { // desynchronization of cartRules, need to be fixed
+                sort($orderCartRulesIds);
```

Comment:
All the sort and array_diff algorithms that have been used so far have had issues There's always a use case that doesn't work with this implementation Besides it's not "developer friendly" as the algorithm is not understandable straight away

I would advise a clearer implementation (from a human perspective) like what has been done in this PR https://github.com/PrestaShop/PrestaShop/pull/19698 in the separated handlers and merge the principle into a generic in this (awesome) service

Basically, a double loop through order cart rules + new order cart rules (I don't think we need the previous ones since they are in order already), from which you populate two arrays `$orderCartRulesToBeRemoved` `$missingOrderCartRules` And then perform the delete/create accordingly